File: 37-sudoku-solver/37-sudoku-solver.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
  const all = Array.from({length: 9}, (_, i) => '' + (i + 1))
  const rows = Array(9).fill(all).map( a => new Set(a))
  const cols = Array(9).fill(all).map( a => new Set(a))
  const boxes = Array(9).fill(all).map( a => new Set(a))
  const unsolved = []
  const rc2b = ([i, j]) => (~~(i/3))*3+(~~(j/3))
  const getValues = ([i, j]) =>
    all.filter(e => rows[i].has(e) && cols[j].has(e) && boxes[rc2b([i, j])].has(e))
  const set = ([i, j], v) => {
    board[i][j] = v
    rows[i].delete(v)
    cols[j].delete(v)
    boxes[rc2b([i, j])].delete(v)
  }
  const unset = ([i, j], v) => {
    rows[i].add(v)
    cols[j].add(v)
    boxes[rc2b([i, j])].add(v)
    board[i][j] = '.'
  }
  const buildContext = () => {
    for (let i = 0; i < 9; i++) {
      for (let j = 0; j < 9; j++) {
        const ch = board[i][j]
        if (ch === '.') {
          unsolved.push([i, j])
        } else {
          rows[i].delete(ch)
          cols[j].delete(ch)
          boxes[rc2b([i, j])].delete(ch)
        }
      }
    }
    unsolved.sort((i0, i1) => getValues(i1).length-getValues(i0).length)
  }
  const solve = () => {
    if (unsolved.length === 0) {
      return true
    }
    if (unsolved.some(i => getValues(i).length === 0)) return false
    const t = unsolved.pop()
    const values = getValues(t)
    for (let v of values) {
      set(t, v)      
      if (solve()) return true
      unset(t, v)
    }
    unsolved.push(t)
    return false
  }
  buildContext()
  solve()  
};

File: 37-sudoku-solver/37-sudoku-solver.cs
public class Solution {
  private bool isValid(char[][] b, int i, int j, char n) {
    foreach (char ch in b[i]) {
      if (ch == n) return false;
    }
    for (int k = 0; k < 9; k++) {
      if (b[k][j] == n) return false;
    }
    int fr = i / 3 * 3;
    int fc = j / 3 * 3;
    for (int r = fr; r < fr + 3; r++) {
      for (int c = fc; c < fc + 3; c++) {
        if (b[r][c] == n) return false;
      }
    }
    return true;
  }
  private bool backtrack(char[][] b, int i, int j) {
    if (j == 9) {
      return backtrack(b, i+1, 0);
    }
    if (i == 9) {
      return true;
    }
    if (b[i][j] != '.') {
      return backtrack(b, i, j+1);
    }
    for (char ch = '1'; ch <= '9'; ch++) {
      if (isValid(b, i, j, ch)) {
        b[i][j] = ch;
        if (backtrack(b, i, j+1))
          return true;
        b[i][j] = '.';
      }
    }
    return false;
  }
  public void SolveSudoku(char[][] board) {
    backtrack(board, 0, 0);
  }
}